File: 1.Elementary level/4.String/Compression/rle/rle.py
def encode_rle(input_str: str) -> str:
    encoded = ""
    count = 1

    for i in range(1, len(input_str) + 1):
        if i < len(input_str) and input_str[i] == input_str[i - 1]:
            count += 1
        else:
            if count == 1:
                encoded += input_str[i - 1]
            elif count == 255:
                encoded += f"255{chr(count)}{input_str[i - 1]}"
            else:
                encoded += f"{count}{input_str[i - 1]}"
            count = 1

    return encoded


def decode_rle(encoded_str: str) -> str:
    decoded = ""
    i = 0

    while i < len(encoded_str):
        if (
            i + 4 < len(encoded_str) and
            encoded_str[i:i+3] == "255"
        ):
            count = ord(encoded_str[i + 3])
            character = encoded_str[i + 4]
            decoded += character * count
            i += 5
        elif encoded_str[i].isdigit():
            count = ""
            while i < len(encoded_str) and encoded_str[i].isdigit():
                count += encoded_str[i]
                i += 1
            character = encoded_str[i]
            decoded += character * int(count)
            i += 1
        else:
            # Handle single characters (no run length)
            decoded += encoded_str[i]
            i += 1

    return decoded


def get_size_in_bits(s: str) -> int:
    return len(s) * 8  # 1 character = 1 byte = 8 bits


# Test the implementation
input_str = "aaabccdddd"
encoded_str = encode_rle(input_str)
decoded_str = decode_rle(encoded_str)

print("Original input:", input_str)
print("Size before compression:", get_size_in_bits(input_str))
print("Encoded output:", encoded_str)
print("Size after compression:", get_size_in_bits(encoded_str))
print("Decoded output:", decoded_str)
print("Size Decode:", get_size_in_bits(decoded_str))
print("======================================")

File: 1.Elementary level/4.String/Compression/rle/rle.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
char* encodeRLE(const char* input) {
    int len = strlen(input);
    char* encoded = (char*)malloc(len * 4); 
    int count = 1, pos = 0;

    for (int i = 1; i <= len; i++) {
        if (i < len && input[i] == input[i - 1]) {
            count++;
        } else {
            if (count == 1) {
                encoded[pos++] = input[i - 1];
            } else if (count == 255) {
                sprintf(&encoded[pos], "255%c%c", count, input[i - 1]);
                pos += 5; 
            } else {
                pos += sprintf(&encoded[pos], "%d%c", count, input[i - 1]);
            }
            count = 1;
        }
    }
    encoded[pos] = '\0';
    return encoded;
}
char* decodeRLE(const char* encoded) {
    int len = strlen(encoded);
    char* decoded = (char*)malloc(len * 4); 
    int pos = 0;

    for (int i = 0; i < len;) {
        if (strncmp(&encoded[i], "255", 3) == 0) {
            int count = (int)encoded[i + 3];
            char ch = encoded[i + 4];
            for (int j = 0; j < count; j++) {
                decoded[pos++] = ch;
            }
            i += 5;
        } else if (isdigit(encoded[i])) {
            int count = 0;
            while (isdigit(encoded[i])) {
                count = count * 10 + (encoded[i] - '0');
                i++;
            }
            char ch = encoded[i];
            for (int j = 0; j < count; j++) {
                decoded[pos++] = ch;
            }
            i++;
        } else {
            decoded[pos++] = encoded[i++];
        }
    }
    decoded[pos] = '\0';
    return decoded;
}

int getSizeInBits(const char* str) {
    return strlen(str) * 8; 
}
int main() {
    const char* input = "aaabccdddd";
    char* encoded = encodeRLE(input);
    char* decoded = decodeRLE(encoded);

    printf("Original input: %s\n", input);
    printf("Size before compression: %d bits\n", getSizeInBits(input));
    printf("Encoded output: %s\n", encoded);
    printf("Size after compression: %d bits\n", getSizeInBits(encoded));
    printf("Decoded output: %s\n", decoded);
    printf("Size after decoding: %d bits\n", getSizeInBits(decoded));

    
    free(encoded);
    free(decoded);

    return 0;
}
